### ISSUE TITLE
cnrm-controller-manager-0 panics from bad IAMPolicy on network, hard to know what's going on

### ISSUE THREAD

**Issue by Scorpiion** [GoogleCloudPlatform/k8s-config-connector#100]
Hi,

I managed to crash the `cnrm-controller-manager-0` process, I traced it back to an error on my end, I create an IAM policy on a network (and you can only do that on subnetworks I think). In a sense I guess it's hard to make it impossible to crash CC, but in this case it was a bit hidden and many people might have a hard time understanding what's is going on. 

So I created this IAM policy ("xxxxxxxxxxxxx-network-1" is a project, but added "x..." for this post):
```
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicy
metadata:
  namespace: xxxxxxxxxxxxx-network-1
  name: xxxxxxxxxxxxx-network-1-network-iampolicy
  annotations:
    cnrm.cloud.google.com/deletion-policy: abandon
spec:
  resourceRef:
    apiVersion: compute.cnrm.cloud.google.com/v1beta1
    kind: ComputeNetwork
    namespace: xxxxxxxxxxxxx-network-1
    name: xxxxxxxxxxxxx-network-1-network
  bindings:
    - members:
        - serviceAccount:cnrm-system@xxxxxxxxxxx.iam.gserviceaccount.com
      role: roles/owner
    - members:
        - user:xxxxxxxxxxxx@example.com
      role: roles/compute.networkUser
```

The key problem above I think is the `spec.resourceRef.kind = ComputeNetwork`, I tried with this one and it also creates the panic:
```
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicy
metadata:
  namespace: xxxxxxxxxxxxx-network-1
  name: bad-network-iampolicy
  annotations:
    cnrm.cloud.google.com/deletion-policy: abandon
spec:
  resourceRef:
    apiVersion: compute.cnrm.cloud.google.com/v1beta1
    kind: ComputeNetwork
    namespace: xxxxxxxxxxxxx-network-1
    name: xxxxxxxxxxxxx-network-1-network
```

`kubectl apply -f ` happily applies both these files.

I often check with `kubectl get` after I created something, so I checked the policy after it was created with, I got this, it looked good so I continued:
```
$ kubectl get iampolicies.iam.cnrm.cloud.google.com --namespace xxxxxxxxxxxxx-network-1 xxxxxxxxxxxxx-network-1-network-iampolicy -o yaml

apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicy
metadata:
  annotations:
    cnrm.cloud.google.com/deletion-policy: abandon
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"iam.cnrm.cloud.google.com/v1beta1","kind":"IAMPolicy","metadata":{"annotations":{"cnrm.cloud.google.com/deletion-policy":"abandon"},"name":"xxxxxxxxxxxxx-network-1-network-iampolicy","namespace":"xxxxxxxxxxxxx-network-1"},"spec":{"bindings":[{"members":["serviceAccount:cnrm-system@r-test-3.iam.gserviceaccount.com"],"role":"roles/owner"},{"members":["user:xxxxxxxxxxxx@example.com"],"role":"roles/compute.networkUser"}],"resourceRef":{"apiVersion":"compute.cnrm.cloud.google.com/v1beta1","kind":"ComputeNetwork","name":"xxxxxxxxxxxxx-network-1-network","namespace":"xxxxxxxxxxxxx-network-1"}}}
  creationTimestamp: "2020-02-26T20:05:26Z"
  generation: 1
  name: xxxxxxxxxxxxx-network-1-network-iampolicy
  namespace: xxxxxxxxxxxxx-network-1
  resourceVersion: "5993643"
  selfLink: /apis/iam.cnrm.cloud.google.com/v1beta1/namespaces/xxxxxxxxxxxxx-network-1/iampolicies/xxxxxxxxxxxxx-network-1-network-iampolicy
  uid: 5443029d-58d3-11ea-9528-4201ac100008
spec:
  bindings:
  - members:
    - serviceAccount:cnrm-system@r-test-3.iam.gserviceaccount.com
    role: roles/owner
  - members:
    - user:xxxxxxxxxxxx@example.com
    role: roles/compute.networkUser
  resourceRef:
    apiVersion: compute.cnrm.cloud.google.com/v1beta1
    kind: ComputeNetwork
    name: xxxxxxxxxxxxx-network-1-network
    namespace: xxxxxxxxxxxxx-network-1
```

What was missing, that I should have picked up on, was that there was no:
```
status:
  conditions:
  - lastTransitionTime: "2020-02-26T10:08:51Z"
    message: The resource is up to date
    reason: UpToDate
    status: "True"
    type: Ready
```

I did not notice that, later when I tried to create some other stuff with CC I noticed that nothing seemed to happen, I checked the logs of CC and saw this stacktrace:
```
{"level":"info","ts":1582746346.1093395,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"spannerinstance-controller","source":"kind source: spanner.cnrm.cloud.google.com/v1beta1, Kind=SpannerInstance"}
{"level":"info","ts":1582746346.1094012,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"computesubnetwork-controller","worker count":20}
{"level":"info","ts":1582746346.1094623,"logger":"computesubnetwork-controller","msg":"starting reconcile","resource":{"namespace":"com-xxxxxxxxxxx-tmp-network-1","name":"com-xxxxxxxxxxx-tmp-network-1-subnet-a"}}
{"level":"info","ts":1582746346.1098762,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"iamcustomrole-controller"}
E0226 19:45:46.333300       1 runtime.go:69] Observed a panic: &errors.errorString{s:"unknown value type: "} (unknown value type: )
/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:679
/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:580
/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:350
/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:217
/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:176
/go/src/cnrm.googlesource.com/cnrm/pkg/controller/iam/policy/iampolicy_controller.go:153
/go/src/cnrm.googlesource.com/cnrm/pkg/controller/iam/policy/iampolicy_controller.go:123
/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1357
panic: unknown value type:  [recovered]
	panic: unknown value type: 

goroutine 5286 [running]:
cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x2b20e00, 0xc0054e0980)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
cnrm.googlesource.com/cnrm/pkg/tf/iamclient.getReferenceFieldValue(0xc0023c7320, 0x5a, 0xc003c5f7a0, 0x0, 0xc0021ab6d0, 0x0, 0x0)
	/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:580 +0x162
cnrm.googlesource.com/cnrm/pkg/tf/iamclient.(*TFIAMClient).buildUnstructuredIAMSkeletonFromReference(0xc00092d380, 0x3617ec0, 0xc001d24ba0, 0x3025ac0, 0xc003d12680, 0xc00201e020, 0x1c, 0xc003a7a1c0, 0xe, 0xc00201e080, ...)
	/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:350 +0x49f
cnrm.googlesource.com/cnrm/pkg/tf/iamclient.(*TFIAMClient).newIAMSkeletonResourceFromReference(0xc00092d380, 0x3617ec0, 0xc001d24ba0, 0x3025ac0, 0xc003d12680, 0xc003a7a1c0, 0xe, 0xc00201e080, 0x1c, 0xc000496630, ...)
	/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:217 +0xae
cnrm.googlesource.com/cnrm/pkg/tf/iamclient.(*TFIAMClient).GetPolicy(0xc00092d380, 0x3617ec0, 0xc001d24ba0, 0xc003a7a1c0, 0xe, 0xc00201e080, 0x1c, 0xc000496630, 0x24, 0xc000496600, ...)
	/go/src/cnrm.googlesource.com/cnrm/pkg/tf/iamclient/tfiamclient.go:176 +0xf4
cnrm.googlesource.com/cnrm/pkg/controller/iam/policy.(*reconcileContext).doReconcile(0xc000d9fc60, 0xc003631d40, 0xc000048220, 0xc0010f0580)
	/go/src/cnrm.googlesource.com/cnrm/pkg/controller/iam/policy/iampolicy_controller.go:153 +0x48f
cnrm.googlesource.com/cnrm/pkg/controller/iam/policy.(*ReconcileIAMPolicy).Reconcile(0xc0020c1d10, 0xc0010f0580, 0x1c, 0xc0013fd7d0, 0x2e, 0x0, 0x0, 0x0, 0x0)
	/go/src/cnrm.googlesource.com/cnrm/pkg/controller/iam/policy/iampolicy_controller.go:123 +0x580
cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0020be280, 0x0)
	/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x20a
cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0004bbb00)
	/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x5e
cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0004bbb00, 0x3b9aca00, 0x0, 0x1, 0xc00056e5a0)
	/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xf8
cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc0004bbb00, 0x3b9aca00, 0xc00056e5a0)
	/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x32e
```

Could we maybe catch this bad config already at the `kubectl apply` stage maybe? I think it would make it a lot easier for people who run into this type of bad config, that I think this will happen a fair amount.

**Comment by kibbles-n-bytes** [comment]
Hey Scorpiion, could you share which version of Config Connector you are running? (i.e., what is the image tag on the `cnrm-controller-manager-0` pod) .

**Comment by kibbles-n-bytes** [comment]
Also, could you clarify whether the pod actually crashed due to the panic, or whether the panic was observed but the pod overall recovered? We added some additional protections against panics in 1.1.0 in order to avoid the case where one bad resource brought down all our controllers.

As for this particular issue with making an IAMPolicy for an unsupported resource, we added a fix for this case in 1.1.0, which is why I wonder if you're on an older version and just need to upgrade.

Agreed also that it is confusing if there are no conditions put on the resource; it's an unintuitive UX to force you to go to the logs. We'll look to reflect this case in the status as well.

**Comment by Scorpiion** [comment]
Hi Michael,

This is the version/tag I'm running:
```
    Image ID:      docker-pullable://gcr.io/cnrm-eap/controller@sha256:039f8c03065f148a500b4c7026a1afe0071e24b513fcaadadde36a5f62910800
```

_"Also, could you clarify whether the pod actually crashed due to the panic, or whether the panic was observed but the pod overall recovered?"_

The pod crashed and went into a restart loop, it did not recover until I removed the bad resource. 

What is the recommended workflow to make sure I'm always is running a good stable version btw? I did not find anything about this in the docs, but maybe I missed it.

**Comment by AlexBulankou** [comment]
> What is the recommended workflow to make sure I'm always is running a good stable version btw? I did not find anything about this in the docs, but maybe I missed it.

@Scorpiion , we are releasing weekly, however today there's no easy way to ensure that you always stay on the latest deployment. To manually integrate it into your workflow you would need to add uninstall and install steps from the latest drop (gs://cnrm/latest/release-bundle.tar.gz)
We are planning improvements to the installation experience later in 2020 and we'll update this thread then.

**Comment by kibbles-n-bytes** [comment]
Thanks Scorpiion; you're currently on version 1.0.1, which didn't have the IAM improvements in place yet. Upgrading should solve this issue.

Note that you don't need to uninstall to upgrade while we're in 1.X; you can just download the latest install bundle for your install mode (i.e. gcp identity, workload identity, or namespaced mode) and apply it on top.

**Comment by kibbles-n-bytes** [comment]
As the explicit issue being addressed here is the panic and crash that was fixed in 1.1.0, I'm going to go ahead and close this issue out. Feel free to open other issues if you have any additional issues, or suggestions for documentation improvements.

**Comment by Scorpiion** [comment]
Hi guys, sorry for my late reply, thanks for the updates, I will update manually for now then. :+1:
